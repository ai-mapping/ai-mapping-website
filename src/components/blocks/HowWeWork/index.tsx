'use client';

import React, { useState } from 'react';
import Image, { StaticImageData } from 'next/image';
import Section from '@/components/layout/Section';

// Add custom CSS for animations
const fadeInStyle = `
  @keyframes fadeIn {
    from { opacity: 0; transform: translateY(10px); }
    to { opacity: 1; transform: translateY(0); }
  }
`;

// Import images
import step1Image from './step1.png';
import step2Image from './step2.png';
import step3Image from './step3.png';
import step4Image from './step4.png';

// Define types for workflow steps
type WorkflowStep = {
  id: number;
  title: string;
  description: string;
  image: StaticImageData;
};

// Main props interface
export type HowWeWorkProps = {
  heading: string;
  description: string;
  steps: WorkflowStep[];
};

// Default props with content
export const defaultHowWeWorkProps: HowWeWorkProps = {
  heading: 'How AI-Mapping Works',
  description:
    'A simple four-step process that transforms your content into fully-functional components.',
  steps: [
    {
      id: 1,
      title: 'Step 1: Write Content in Markdown',
      description:
        'Create your content in markdown files with simple style annotations. No need to worry about component structure or technical implementation details - just focus on your message.',
      image: step1Image,
    },
    {
      id: 2,
      title: 'Step 2: AI Analyzes Your Content',
      description:
        'AI-Mapping analyzes your content structure, intent, and styling annotations to determine the optimal component structure for presenting your information.',
      image: step2Image,
    },
    {
      id: 3,
      title: 'Step 3: Components Are Generated',
      description:
        'The system automatically maps content to existing components or generates new ones as needed, all while respecting your design system and guidelines.',
      image: step3Image,
    },
    {
      id: 4,
      title: 'Step 4: Review and Refine',
      description:
        'Developers can review, adjust, and refine the generated components to ensure they meet technical requirements and best practices before finalizing the implementation.',
      image: step4Image,
    },
  ],
};

const HowWeWork = ({
  heading = defaultHowWeWorkProps.heading,
  steps = defaultHowWeWorkProps.steps,
}: Partial<HowWeWorkProps>) => {
  // State to track which step is selected
  const [selectedStepId, setSelectedStepId] = useState<number>(1);

  // Handle step click
  const handleStepClick = (id: number) => {
    console.log('Clicked step:', id, 'Previous step:', selectedStepId);
    setSelectedStepId(id);
  };

  // Get the selected step
  const selectedStep =
    steps.find((step) => step.id === selectedStepId) || steps[0];

  return (
    <Section id="how-we-work" background="surface" paddingY="lg" containerWidth="container">
      <style dangerouslySetInnerHTML={{ __html: fadeInStyle }} />
      <h2 className="text-2xl md:text-3xl font-semibold text-center mb-10 text-white">
        {heading}
      </h2>

        <div className="grid grid-cols-1 md:grid-cols-4 gap-6 mb-10">
          {steps.map((step) => (
            <div
              key={`step-${step.id}`}
              className={`cursor-pointer transition-all duration-300 ${
                selectedStepId === step.id
                  ? 'border-2 border-primary scale-105 shadow-lg bg-primary/10'
                  : 'border border-neutral-dark hover:border-primary hover:shadow-md bg-neutral-light/20'
              } overflow-hidden rounded-lg`}
              role="button"
              aria-pressed={selectedStepId === step.id}
              tabIndex={0}
              // onKeyDown={(e) => e.key === 'Enter' && handleStepClick(step.id)}
              onClick={() => handleStepClick(step.id)}
            >
              <div className="p-0">
                <Image
                  src={step.image}
                  alt={step.title}
                  className="w-full h-auto"
                  width={300}
                  height={400}
                  priority={step.id === 1}
                />
              </div>
              <div className="p-4 bg-neutral-light h-24 border-t border-neutral-dark">
                <h3 className="font-medium text-center text-white">{step.title}</h3>
              </div>
            </div>
          ))}
        </div>

        <div
          key={`description-${selectedStepId}`}
          className="bg-neutral-light rounded-lg p-8 transition-all duration-300 border-2 border-primary shadow-lg"
          style={{ animation: 'fadeIn 0.5s ease-in-out' }}
        >
          <h3 className="text-xl font-semibold mb-4 text-primary">
            {selectedStep.title}
          </h3>
          <p className="text-neutral">{selectedStep.description}</p>
          <div className="text-right mt-4 text-sm text-primary">
            Step {selectedStepId} of {steps.length}
          </div>
        </div>
    </Section>
  );
};

export default HowWeWork;