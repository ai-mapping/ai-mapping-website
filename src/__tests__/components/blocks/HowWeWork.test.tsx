import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import HowWeWork, { defaultHowWeWorkProps } from '@/components/blocks/HowWeWork';
import { StaticImageData } from 'next/image';

// Mock the step images
jest.mock('@/components/blocks/HowWeWork/step1.png', () => ({
  src: '/step1.png',
  height: 400,
  width: 300
}));

jest.mock('@/components/blocks/HowWeWork/step2.png', () => ({
  src: '/step2.png',
  height: 400,
  width: 300
}));

jest.mock('@/components/blocks/HowWeWork/step3.png', () => ({
  src: '/step3.png',
  height: 400,
  width: 300
}));

jest.mock('@/components/blocks/HowWeWork/step4.png', () => ({
  src: '/step4.png',
  height: 400,
  width: 300
}));

describe('HowWeWork Component', () => {
  it('renders with default props correctly', () => {
    const { container } = render(<HowWeWork />);
    
    // Check if heading is rendered
    expect(screen.getByText(defaultHowWeWorkProps.heading)).toBeInTheDocument();
    
    // Check if all steps are rendered with specific queries
    defaultHowWeWorkProps.steps.forEach((step) => {
      const stepElements = screen.getAllByText(step.title);
      expect(stepElements.length).toBeGreaterThan(0);
    });
    
    // Check if step 1 is selected by default (as per the useState(1) in the component)
    const stepButtons = screen.getAllByRole('button');
    const step1 = stepButtons.find(button => button.getAttribute('aria-pressed') === 'true');
    expect(step1).toBeInTheDocument();
    expect(step1).toHaveAttribute('aria-pressed', 'true');
    
    // Check if step 1 description is shown
    expect(screen.getByText(defaultHowWeWorkProps.steps[0].description)).toBeInTheDocument();
    
    // Check if step counter is shown - we need to use a regex for the updated text
    expect(screen.getByText(/Step 1 of/)).toBeInTheDocument();
    
    // Snapshot test
    expect(container).toMatchSnapshot();
  });
  
  it('changes selected step when clicking on a step', () => {
    render(<HowWeWork />);
    
    // Initially step 1 is selected
    expect(screen.getByText(/Step 1 of/)).toBeInTheDocument();
    
    // Get all buttons and find step 4
    const buttons = screen.getAllByRole('button');
    // Find the button that contains step 4 title text
    const step4 = buttons.find(button => 
      button.textContent.includes(defaultHowWeWorkProps.steps[3].title)
    );
    
    // Click on step 4
    fireEvent.click(step4);
    
    // Check if step 4 is now selected
    expect(step4).toHaveAttribute('aria-pressed', 'true');
    
    // Check if step 4 description is shown
    expect(screen.getByText(defaultHowWeWorkProps.steps[3].description)).toBeInTheDocument();
    
    // Check if step counter is updated
    expect(screen.getByText(/Step 4 of/)).toBeInTheDocument();
  });
  
  it('renders with custom props correctly', () => {
    const customProps = {
      heading: 'Custom How We Work Heading',
      steps: [
        {
          id: 1,
          title: 'Custom Step 1',
          description: 'Custom Step 1 Description',
          image: { src: '/custom-step1.png', height: 300, width: 200 } as StaticImageData
        },
        {
          id: 2,
          title: 'Custom Step 2',
          description: 'Custom Step 2 Description',
          image: { src: '/custom-step2.png', height: 300, width: 200 } as StaticImageData
        }
      ]
    };
    
    const { container } = render(<HowWeWork {...customProps} />);
    
    // Check if custom heading is rendered
    expect(screen.getByText(customProps.heading)).toBeInTheDocument();
    
    // Check if custom steps are rendered
    customProps.steps.forEach((step) => {
      const stepElements = screen.getAllByText(step.title);
      expect(stepElements.length).toBeGreaterThan(0);
    });
    
    // Step 1 should be selected by default (as per the useState(1) in the component)
    const stepButtons = screen.getAllByRole('button');
    const step1 = stepButtons.find(button => button.getAttribute('aria-pressed') === 'true');
    expect(step1).toBeInTheDocument();
    expect(step1).toHaveAttribute('aria-pressed', 'true');
    
    // Check if step 1 description is shown
    expect(screen.getByText(customProps.steps[0].description)).toBeInTheDocument();
    
    // Check if step counter is shown
    expect(screen.getByText(/Step 1 of/)).toBeInTheDocument();
    
    // Snapshot test
    expect(container).toMatchSnapshot();
  });
});