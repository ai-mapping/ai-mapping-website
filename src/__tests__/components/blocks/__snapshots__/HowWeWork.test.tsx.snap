// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HowWeWork Component renders with custom props correctly 1`] = `
<div>
  <section
    class="bg-surface py-12"
    id="how-we-work"
  >
    <div
      class="container mx-auto px-4 md:px-6 lg:px-8"
    >
      <style>
        
  @keyframes fadeIn {
    from { opacity: 0; transform: translateY(10px); }
    to { opacity: 1; transform: translateY(0); }
  }

      </style>
      <h2
        class="text-2xl md:text-3xl font-semibold text-center mb-10 text-white"
      >
        Custom How We Work Heading
      </h2>
      <div
        class="grid grid-cols-1 md:grid-cols-4 gap-6 mb-10"
      >
        <div
          aria-pressed="true"
          class="cursor-pointer transition-all duration-300 border-2 border-primary scale-105 shadow-lg bg-primary/10 overflow-hidden rounded-lg"
          role="button"
          tabindex="0"
        >
          <div
            class="p-0"
          >
            <img
              alt="Custom Step 1"
              class="w-full h-auto"
              data-testid="next-image-Custom Step 1"
              height="400"
              src="/custom-step1.png"
              width="300"
            />
          </div>
          <div
            class="p-4 bg-neutral-light h-24 border-t border-neutral-dark"
          >
            <h3
              class="font-medium text-center text-white"
            >
              Custom Step 1
            </h3>
          </div>
        </div>
        <div
          aria-pressed="false"
          class="cursor-pointer transition-all duration-300 border border-neutral-dark hover:border-primary hover:shadow-md bg-neutral-light/20 overflow-hidden rounded-lg"
          role="button"
          tabindex="0"
        >
          <div
            class="p-0"
          >
            <img
              alt="Custom Step 2"
              class="w-full h-auto"
              data-testid="next-image-Custom Step 2"
              height="400"
              src="/custom-step2.png"
              width="300"
            />
          </div>
          <div
            class="p-4 bg-neutral-light h-24 border-t border-neutral-dark"
          >
            <h3
              class="font-medium text-center text-white"
            >
              Custom Step 2
            </h3>
          </div>
        </div>
      </div>
      <div
        class="bg-neutral-light rounded-lg p-8 transition-all duration-300 border-2 border-primary shadow-lg"
        style="animation: fadeIn 0.5s ease-in-out;"
      >
        <h3
          class="text-xl font-semibold mb-4 text-primary"
        >
          Custom Step 1
        </h3>
        <p
          class="text-neutral"
        >
          Custom Step 1 Description
        </p>
        <div
          class="text-right mt-4 text-sm text-primary"
        >
          Step 
          1
           of 
          2
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`HowWeWork Component renders with default props correctly 1`] = `
<div>
  <section
    class="bg-surface py-12"
    id="how-we-work"
  >
    <div
      class="container mx-auto px-4 md:px-6 lg:px-8"
    >
      <style>
        
  @keyframes fadeIn {
    from { opacity: 0; transform: translateY(10px); }
    to { opacity: 1; transform: translateY(0); }
  }

      </style>
      <h2
        class="text-2xl md:text-3xl font-semibold text-center mb-10 text-white"
      >
        How AI-Mapping Works
      </h2>
      <div
        class="grid grid-cols-1 md:grid-cols-4 gap-6 mb-10"
      >
        <div
          aria-pressed="true"
          class="cursor-pointer transition-all duration-300 border-2 border-primary scale-105 shadow-lg bg-primary/10 overflow-hidden rounded-lg"
          role="button"
          tabindex="0"
        >
          <div
            class="p-0"
          >
            <img
              alt="Step 1: Write Content in Markdown"
              class="w-full h-auto"
              data-testid="next-image-Step 1: Write Content in Markdown"
              height="400"
              src="/step4.png"
              width="300"
            />
          </div>
          <div
            class="p-4 bg-neutral-light h-24 border-t border-neutral-dark"
          >
            <h3
              class="font-medium text-center text-white"
            >
              Step 1: Write Content in Markdown
            </h3>
          </div>
        </div>
        <div
          aria-pressed="false"
          class="cursor-pointer transition-all duration-300 border border-neutral-dark hover:border-primary hover:shadow-md bg-neutral-light/20 overflow-hidden rounded-lg"
          role="button"
          tabindex="0"
        >
          <div
            class="p-0"
          >
            <img
              alt="Step 2: AI Analyzes Your Content"
              class="w-full h-auto"
              data-testid="next-image-Step 2: AI Analyzes Your Content"
              height="400"
              src="/step4.png"
              width="300"
            />
          </div>
          <div
            class="p-4 bg-neutral-light h-24 border-t border-neutral-dark"
          >
            <h3
              class="font-medium text-center text-white"
            >
              Step 2: AI Analyzes Your Content
            </h3>
          </div>
        </div>
        <div
          aria-pressed="false"
          class="cursor-pointer transition-all duration-300 border border-neutral-dark hover:border-primary hover:shadow-md bg-neutral-light/20 overflow-hidden rounded-lg"
          role="button"
          tabindex="0"
        >
          <div
            class="p-0"
          >
            <img
              alt="Step 3: Components Are Generated"
              class="w-full h-auto"
              data-testid="next-image-Step 3: Components Are Generated"
              height="400"
              src="/step4.png"
              width="300"
            />
          </div>
          <div
            class="p-4 bg-neutral-light h-24 border-t border-neutral-dark"
          >
            <h3
              class="font-medium text-center text-white"
            >
              Step 3: Components Are Generated
            </h3>
          </div>
        </div>
        <div
          aria-pressed="false"
          class="cursor-pointer transition-all duration-300 border border-neutral-dark hover:border-primary hover:shadow-md bg-neutral-light/20 overflow-hidden rounded-lg"
          role="button"
          tabindex="0"
        >
          <div
            class="p-0"
          >
            <img
              alt="Step 4: Review and Refine"
              class="w-full h-auto"
              data-testid="next-image-Step 4: Review and Refine"
              height="400"
              src="/step4.png"
              width="300"
            />
          </div>
          <div
            class="p-4 bg-neutral-light h-24 border-t border-neutral-dark"
          >
            <h3
              class="font-medium text-center text-white"
            >
              Step 4: Review and Refine
            </h3>
          </div>
        </div>
      </div>
      <div
        class="bg-neutral-light rounded-lg p-8 transition-all duration-300 border-2 border-primary shadow-lg"
        style="animation: fadeIn 0.5s ease-in-out;"
      >
        <h3
          class="text-xl font-semibold mb-4 text-primary"
        >
          Step 1: Write Content in Markdown
        </h3>
        <p
          class="text-neutral"
        >
          Create your content in markdown files with simple style annotations. No need to worry about component structure or technical implementation details - just focus on your message.
        </p>
        <div
          class="text-right mt-4 text-sm text-primary"
        >
          Step 
          1
           of 
          4
        </div>
      </div>
    </div>
  </section>
</div>
`;
